Userland/Libraries/LibWeb/Layout/FlexFormattingContext.cpp

```diff
@@ -19,49 +24,666 @@ FlexFormattingContext::~FlexFormattingContext()
+            if (flex_item.max_content_flex_fraction > 0) {
+                product = largest_max_content_flex_fraction * flex_item.box.computed_values().flex_grow_factor().value_or(1);
+            } else {
+                product = largest_max_content_flex_fraction * max(flex_item.box.computed_values().flex_shrink_factor().value_or(1), 1.0f) * flex_item.flex_base_size;
+            }
+            result += flex_item.flex_base_size + product;
+        }
+        main_available_size = clamp(result, main_min_size, main_max_size);
+    }
+    set_main_size(box, main_available_size);
+
+    /* 5. Collect flex items into flex lines: */
+    Vector<FlexLine> flex_lines {};
+    // FIXME: Also support wrap-reverse
+    if (box.computed_values().flex_wrap() == CSS::FlexWrap::Nowrap) {
+        FlexLine line = {};
+        for (auto& flex_item : flex_items) {
+            line.items.append(&flex_item);
+        }
+        flex_lines.append(line);
+    } else {
+        FlexLine line = {};
+        float line_main_size = 0;
+        for (auto& flex_item : flex_items) {
+            if ((line_main_size + flex_item.hypothetical_main_size) > main_available_size) {
+                flex_lines.append(line);
+                line = {};
+                line_main_size = 0;
+            }
+            line.items.append(&flex_item);
+            line_main_size += flex_item.hypothetical_main_size;
+        }
+        flex_lines.append(line);
+    }
+
+    /* 6. Resolve the flexible lengths */
+    enum FlexFactor {
+        FlexGrowFactor,
+        FlexShrinkFactor
+    };
+
+    FlexFactor used_flex_factor;
+    /* 6.1. Determine used flex factor */
+    for (auto& flex_line : flex_lines) {
+
+        size_t number_of_unfrozen_items_on_line = flex_line.items.size();
+
+        float sum_of_hypothetical_main_sizes = 0;
+        for (auto& flex_item : flex_line.items) {
+            sum_of_hypothetical_main_sizes += flex_item->hypothetical_main_size;
+        }
+        if (sum_of_hypothetical_main_sizes < main_available_size)
+            used_flex_factor = FlexFactor::FlexGrowFactor;
         else
-            y += child_box.margin_box_height();
-    });
+            used_flex_factor = FlexFactor::FlexShrinkFactor;
 
-    // Then, compute the height of the entire flex container.
-    // FIXME: This is not correct height calculation..
-    if (horizontal) {
-        box.set_height(tallest_child_height);
-        box.set_width(x);
+        for (auto& flex_item : flex_line.items) {
+            if (used_flex_factor == FlexFactor::FlexGrowFactor)
+                flex_item->flex_factor = flex_item->box.computed_values().flex_grow_factor();
+            else if (used_flex_factor == FlexFactor::FlexShrinkFactor)
+                flex_item->flex_factor = flex_item->box.computed_values().flex_shrink_factor();
+        }
+
+        /* 6.2. Size inflexible items */
+        auto freeze_item_setting_target_main_size_to_hypothetical_main_size = [&](FlexItem* item) {
```

Comment:
This doesn't look like it can handle null items, maybe it can take a reference instead?

Reply:
It seems like this whole pointer thing is here because `Vector` can't store references, perhaps it can be extended to support references (or more specifically, pretend to support references)?

Reply:
I updated it to take a reference but now I dereference the flex_item for the calls to this...
We can be very sure that there won't be any null items there since we control everything that is added to the Vector and we only ever add `FlexItem`s

Reply:
I think it's worth the trouble of derefing it at the very start, makes the rest of the code much less "is this null"-y.

My comment wasn't really about derefing the pointers you get from the vector, but rather changing vector to hand out references (i.e. allowing `Vector<T&>`, and making it hand out references).
It's not really a requirement for this PR, we can do that in another PR.

Reply:
Ah, that's indeed a good thing to do!

Yea, I think that's wayy out of scope for this PR...